### PR TITLE
Add port to macOS

### PR DESCRIPTION
**Pull request by mockbutler**
The macOS porting layer was synthesized taking (and in a few cases modifying) code from both the Linux and Win32 support layers. There is really no "new code" per se.
This necessarily has resulted in duplicate code. However, extracting a common porting subset for just two targets did not seem warranted. The amount of code is small, and the clear boundaries between porting layers aids comprehension. 

Tested and verified working on 10.3 and 10.4 macOS versions with a lightning to Ethernet adapter. 

I have submitted a completed contributor license agreement to the email address as specified.

### CODE REVIEW

**Comment by nakarlsson** [comment]
squash into one macos port commit
What test have you run?

**Comment by mockbutler** [comment]
I've squashed into a single commit, adding detail to the commit message.

I shouldn't call it testing. I have run the port successfully driven by custom software, with both custom tools, and tools like `slaveinfo` against real hardware, both custom and name brand, in a narrow task specific manner, bus speed 100Hz with very little SDO traffic. 

Enough to show that the port works in principal. As the port primarily remixes existing code, I'm moderately confident it should be generally functional, with the exception of `realtime` support.

File: CMakeLists.txt [hunk 1]
```diff
@@ -22,10 +22,15 @@ if(WIN32)
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} /D _CRT_SECURE_NO_WARNINGS")
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS}  /WX")
   set(OS_LIBS wpcap.lib Packet.lib Ws2_32.lib Winmm.lib)
-elseif(UNIX)
+elseif(UNIX AND NOT APPLE)
   set(OS "linux")
   set(CMAKE_C_FLAGS "${CMAKE_C_FLAGS} -Wall -Wextra -Werror")
   set(OS_LIBS pthread rt)
+elseif(APPLE)
+  # This must come *before* linux or MacOSX will identify as Unix.
```
[the remark this answers is not in the document]
Comment (mockbutler) [reply CMakeLists.txt]: Removed

**Comment by mockbutler** [comment]
Rather than fix the spelling mistakes, I've reworked the port to create a common unix port from the most recent linux files, with specialization for linux and macos in oshw. This should minimize the cost of future maintenance.

I have also changed the port name to "macos". I don't understand why the Apple keep kicking the abbreviated name for the OS around, but that is the name used in Wikipedia.

I have only had a chance to test the build on macos. I'm assuming the CI will go red if the Linux build fails.

**Comment by nakarlsson** [comment]
I prefer if you initially keep Linux and macOS/Unix separate. Merging them can be a future improvement. So, please stick with the original PR but start from the latest Linux files

**Comment by cklosters** [comment]
Any news on this PR? We're looking into supporting CoE in our engine but need OSX support next to Win64 and Linux.

**Comment by mockbutler** [comment]
I've restored the separate files for the macosx port as requested.

**Pull request merged**